### ISSUE TITLE
Simplifying parameterizations 

### ISSUE THREAD

**Issue by josephdviviano** [saleml/torchgfn#77]
OK so I think we can de-complexify dramatically the Parameterization - Estimator framework.

Right now, `Parameterizations` (which are tied to specific losses) accept multiple `Estimators`, which in turn might contain `nn.Module`s or `torch.Tensor`s.

This is tricky for a few reasons. First we need to do a tonne of bookkeeping when we are doing things like saving/loading parameters, or calling `Paramaterization.named_parameters()` - in order to actually return the correct dict of parameters, the `Paramaterization` and `Estimator` classes need to handle everything correctly and store the correct dict manually. This should not be required.

This is also tricky when using optimizers, because you need to feed the right subset of parameters to the optimizer. Sometimes there's a trainable logz parameter, but it won't be in `named_parameters()`, and you need to point the optimizer at it explicitly. If you don't do this the model won't learn well but you will never know why without serious digging.

I would propose that I re-write these classes using inheritance and mixin the `nn.Module` class directly such that pytorch features like `parameters()` and `save()` **just work**. The reason I'm concerned here is this is an excellent vector through which to build in a silent bug. In fact I think it's highly likely this will happen when we start getting external users. 

The expected behavior is that feeding `Parameterization.parameters()` to Adam would just work. Or saving would just work - everything would get saved and loaded and without any concern that the user must handle edge cases or other logic manually.

Thoughts?

**Comment by saleml** [comment]
I'm in favor for such a change. Two things:
- Will `Parameterization.parameters()` handle parameter tying / sharing so that no duplicates are passed to optimizers?
- Shall we change the name `parametrization` altogether, and just call it... `gflownet`, now that `loss` is a method of it ?

**Comment by josephdviviano** [comment]
> Will `Parameterization.parameters()` handle parameter tying / sharing so that no duplicates are passed to optimizers?

Yes that would be one major benefit. Seeing as it seems you might want a different optimizer for `logZ`, and potentially for the `P_f` and `P_b`, they should be accessible independently as well, but the default `parameters()` should give you everything.

> Shall we change the name parametrization altogether, and just call it... gflownet, now that loss is a method of it ?

I think so. We should come up with a nomenclature and structure today we're both happy with on the board. A user will typically just instantiate a `GFlowNet` (such as TB) with all the pieces pre-configured. Otherwise, they would build their own with a series of `Estimators` (or `Modules)`, a loss, etc. But crucially, a `GFlowNet` will always be a valid Pytorch `nn.Module`.

**Comment by josephdviviano** [comment]
Just FYI this work is ongoing but proving to be quite a tangle, I might need to run by some design decisions with the group post Hawaii party

**Comment by saleml** [comment]
Handled in #89
